src/middlewared/middlewared/plugins/idmap.py

```diff
@@ -0,0 +1,1103 @@
+            domain_info = await self.middleware.call(
+                'idmap.domain.query', [('domain', '=', data['domain']['idmap_domain_dns_domain_name'].upper())]
+            )
+            data['domain'] = domain_info[0]['id']
+
+        return data
+
+    @private
+    async def _common_validate(self, data):
+        """
+        Common validation checks for all idmap backends.
+        """
+        verrors = ValidationErrors()
+        if data['range_high'] < data['range_low']:
+            verrors.add(f'idmap_{idmap_type}.update', 'Idmap high range must be greater than idmap low range')
+            return verrors
+            
+        configured_domains = await self.get_configured_idmap_domains()
+        new_range = range(data['range_low'], data['range_high'])
+        for i in configured_domains:
+            if i['domain']['idmap_domain_name'] == data['domain']['idmap_domain_name']:
+                continue
+            existing_range = range(i['backend_data']['range_low'], i['backend_data']['range_high'])
+            if range(max(existing_range[0], new_range[0]), min(existing_range[-1], new_range[-1])+1):
+                verrors.add(
+                    f'idmap_range', 
+                    f'new idmap range conflicts with existing range for domain [{i["domain"]["idmap_domain_name"]}]'
+                )
+
+        return verrors
+        
+
+    @accepts()
+    async def get_configured_idmap_domains(self):
+        """
+        Inner join domain-to-backend table with its corresponding idmap backend
+        table on the configured short-form domain name. Sorted by idmap high range.
+        """
+        domains =  await self.middleware.call('idmap.domaintobackend.query')
+        configured_domains = []
+        for domain in domains:
+            b = await self.middleware.call(
+                f'idmap.{domain["idmap_backend"]}.query',
+                [('domain', '=', domain['domain']['idmap_domain_name'])]
+            )
+            for entry in b:
+                entry.pop('domain')
+                entry.pop('id')
+                domain.update({'backend_data': entry})
+                configured_domains.append(domain)
+
+        return sorted(configured_domains, key=lambda domain: domain['backend_data']['range_high'])
+
+    @private
+    async def get_next_idmap_range(self):
+        """
+        Increment next high range by 100,000,000 ids. This number has
+        to accomodate the highest available rid value for a domain.
+        Configured idmap ranges _must_ not overlap.
+        """
+        sorted_idmaps = await self.get_configured_idmap_domains()
+        low_range = sorted_idmaps[-1]['backend_data']['range_high'] + 1
+        high_range = sorted_idmaps[-1]['backend_data']['range_high'] + 100000000
+        return (low_range, high_range)
+
+    @accepts()
+    async def clear_idmap_cache(self):
+        await self.middleware.call('service.stop', 'smb')
+        os.remove('/var/db/samba4/winbindd_cache.tdb')
```

Comment:
Perhaps its safer to have a try/except here?

Reply:
I agree.